### PR TITLE
development

### PR DESCRIPTION
**Pull request by asmitzz**
please review my ecommerce-site

### CODE REVIEW

**Comment by sneha-pujari** [comment]
Home button/link to the products page can be added to the cart page, since after going to the cart I couldn't easily navigate my way back to add products again.
Everything else, looks good.

**Comment by asmitzz** [comment]
> Home button/link to the products page can be added to the cart page, since after going to the cart I couldn't easily navigate my way back to add products again.
> Everything else, looks good.

For that we will use router which we will be going to learn on Saturday ,so that you can easily navigate back to the products . For now you can click on the shoppingHub in the navbar to go back to the products page. For cart click on cart icon and so on.

**Comment by jugalkpatel** [comment]
Please Add your live link on about.

**Comment by asmitzz** [comment]
> Please Add your live link on about.

Done

File: client/src/pages/Wishlist/WishList.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      <div className="wishlist-container">
+          <Toast show={error} onClick={() => setError(false)} message="Something went wrong with server" error={true} background="red" color="white"/>
+
+      <Spinner show={spinner}/>
+        <h1 style={{textAlign: 'center'}}>Wishlist ({wishlist.length})</h1>
+        <div style={{display: wishlist.length > 0 ? 'grid': "block"}}  className="products">
+              {
+                 wishlist.length > 0 ? wishlist.map( item => (
+                    <div key={item._id} className={ item.stock ? "card" : "card out-of-stock"}>
+                      <div className="remove-icon">
```
Comment (jugalkpatel) [inline client/src/pages/Wishlist/WishList.js:28]: fix remove icon. seee this
https://shareit.video/ItBWfU6AxOv1yzTNCNgo

File: server/controllers/carts.controllers.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
+const Carts = require('../models/cart.model');
+const Products = require('../models/product.model');
+const Users = require('../models/user.model');
+
+const checkCart = async(req,res,next,uid) => {
+    try {
+        const cart = await Carts.findOne({uid});
+        req.cart = cart;
```
Comment (jugalkpatel) [inline server/controllers/carts.controllers.js:8]: Sugesstion: === would be good here.

File: server/controllers/carts.controllers.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          return res.status(200).json({cart:products})
+        }
+        res.status(200).json({cart:[]})
+    } catch (error) {
+         res.status(404).json({ message:"Cart not found" })
+     }
+
+}
+
+const addProductInCart = async(req,res) => {
```
Comment (jugalkpatel) [inline server/controllers/carts.controllers.js:46]: code is not in the try-catch block

File: server/controllers/carts.controllers.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    const createCart = await Carts.create({ uid,products:[{product:product.id,quantity}] });
+    let userReference = await Users.findById(uid);
+    userReference.cart = createCart.id;
+    userReference = await userReference.save()
+    res.status(200).json({ message:"Product added in cart"})
+}
+
+const removeProductFromCart = async(req,res) => {
```
Comment (jugalkpatel) [inline server/controllers/carts.controllers.js:82]: code is not in the try-catch block. parameters from body can be undefined here.

File: server/controllers/carts.controllers.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    let { cart,product } = req;
+
+    await product.remove()
+    await cart.save();
+    res.status(200).json({message:"Product removed from cart"})
+}
+
+const increaseQuantityOfProduct = async(req,res) => {
```
Comment (jugalkpatel) [inline server/controllers/carts.controllers.js:97]: code is not in the try-catch block. parameters from body can be undefined here

File: server/controllers/carts.controllers.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    let { cart,product } = req;
+
+    product.quantity++
+    await cart.save();
+    res.status(200).json({message:"Product quantity increased"})
+}
+
+const decreaseQuantityOfProduct = async(req,res) => {
```
Comment (jugalkpatel) [inline server/controllers/carts.controllers.js:112]: code is not in the try-catch block. parameters from the body can be undefined here.

File: server/controllers/carts.controllers.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    let { cart,product } = req;
+
+    product.quantity--;
+    await cart.save();
+    res.status(200).json({message:"Product quantity decreased"})
+}
+
+const emptyCart = async(req,res) => {
```
Comment (jugalkpatel) [inline server/controllers/carts.controllers.js:90]: code is not in the try-catch block. parameters from body can be undefined here.

File: client/src/pages/Wishlist/WishList.js (hunk 1 above)
Comment (asmitzz) [reply client/src/pages/Wishlist/WishList.js:28]: fixed

File: client/src/reducers/CartReducer.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+const CartReducer = () => {
+
+    const navigate = useNavigate();
+    const { isUserloggedIn,uid } = useAuth();
+
+    useEffect( () => {
+      ( async function(){
+        try {
+          const res = await axios.get("https://shopping-hub-2021.herokuapp.com/api/carts/"+uid);
```
Comment (jugalkpatel) [inline client/src/reducers/CartReducer.js:18]: Suggestion: template literal could be more reablable.

File: client/src/pages/yourorders/YourOrders.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+const MyOrders = () => {
+    const {uid} = useAuth();
+    const [orders,setOrders] = useState([]);
+    const [spinner,setSpinner] = useState(false);
+
+    useEffect( () => {
+        ( async function(){
+          setSpinner(true)
+          try {
+            const res = await axios.get("https://shopping-hub-2021.herokuapp.com/api/orders/"+uid);
```
Comment (jugalkpatel) [inline client/src/pages/yourorders/YourOrders.js:17]: Suggestion: template literal would be more readable.

**Comment by jugalkpatel** [comment]
Suggestions: 
- add limit in an address in address management.
- total number of items in the products list.
- 201 can be used as response while adding records to the DB.
- change your dispatch calls according to responses over time.
- make your code formatting better.

File: client/src/App.css [hunk 10]
```diff
@@ -573,6 +576,10 @@ fieldset input {
     margin-left: 7%;
   }
 
+  .filter-content-label{
+    font-size:0.8rem
+  }
+
   @media(max-width:800px){
```
Comment (Ayush-kumar1) [inline client/src/App.css]: Suggestion: Try to make separate CSS files for separate components. This way it is easier to manipulate them later

File: client/src/reducers/CartReducer.js [hunk 11]
```diff
@@ -10,30 +10,23 @@ const CartReducer = () => {
     const { isUserloggedIn,uid } = useAuth();
 
     useEffect( () => {
+      if(!isUserloggedIn){
+        return;
+      }
       ( async function(){
         try {
           const res = await axios.get("https://shopping-hub-2021.herokuapp.com/api/carts/"+uid);
```
Comment (Ayush-kumar1) [inline client/src/reducers/CartReducer.js:18]: Use template literals here,it's the better approach

**Review by Ayush-kumar1 — COMMENTED** [review]
The code looks good. Few suggestions-
You can use try-catch block at few functions you have missed.
Try to write CSS code for separate components in different css files as this will be easier to manipulate in the future.
Since your app is almost ready you should use relevant image and titles.
Apart from this your app is very good

File: server/controllers/carts.controllers.js (hunk 3 above)
Comment (asmitzz) [reply server/controllers/carts.controllers.js:46]: no we don't need try catch there

File: client/src/pages/yourorders/YourOrders.js (hunk 9 above)
Comment (asmitzz) [reply client/src/pages/yourorders/YourOrders.js:17]: yeah will add templete literals and refactored async calls code later

**Comment by asmitzz** [comment]
> Suggestions:
> 
> * add limit in an address in address management.
> * total number of items in the products list.
> * 201 can be used as response while adding records to the DB.
> * change your dispatch calls according to responses over time.
> * make your code formatting better.



> Suggestions:
> 
> * add limit in an address in address management.
> * total number of items in the products list.
> * 201 can be used as response while adding records to the DB.
> * change your dispatch calls according to responses over time.
> * make your code formatting better.

1. No I don't address limit will be good there.
2. Yeah right it makes sense will add it soon
3. yeah I will work on it and put the right status codes
4. yeah I fixed that now dispatch calls will happen according to responses
5. yeah sure will refactored all the codes

File: client/src/reducers/CartReducer.js (hunk 11 above)
Comment (asmitzz) [reply client/src/reducers/CartReducer.js:18]: yeah trying to figure it out working on making a template literal for async calls

File: client/public/index.html [hunk 12]
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
```
Comment (thesudeshdas) [inline client/public/index.html:1]: Add a favicon, It would look much better. Use Figma for designing a simple logo.

File: client/src/App.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
+import React from "react";
+
+import { Routes,Route } from 'react-router-dom';
```
Comment (thesudeshdas) [inline client/src/App.js:3]: Use a formatter for indentation. I prefer ESlint + Prettier. But ESlint can mess up the code, so beware about that. Here  are the docs - [ESLint](https://eslint.org/docs/user-guide/getting-started) & [Prettier](https://prettier.io/docs/en/index.html).


Check especially this part for prettier [plugin](https://prettier.io/docs/en/integrating-with-linters.html) if you're planning on using both ESLint & Prettier

File: client/src/App.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import DisplayAddresses from "./pages/address/DisplayAddresses";
+import EditAddress from "./pages/address/components/EditAddress";
+import AddNewAddress from "./pages/address/components/AddNewAddress";
+import OrderSummary from "./pages/orderSummary/OrderSummary";
+
+import { useAuth } from "./contexts/AuthContext";
+
+import MyAccount from "./pages/myaccount/MyAccount";
+import YourOrders from "./pages/yourorders/YourOrders";
+// import ForgetPassword from "./auth/ForgetPassword";
```
Comment (thesudeshdas) [inline client/src/App.js:23]: Remove redudant comments. It doesn't make sense to have them here anymore.

File: client/src/api/fakeAuthApi.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
+export const login = (email, password, users) => {
+
+    const findUser = (email) => {
+        return users.find( user => user.email === email)
```
Comment (thesudeshdas) [inline client/src/api/fakeAuthApi.js:4]: Removing return and curly braces will work just fine here.You should remove them

File: client/src/api/fakeAuthApi.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
+export const login = (email, password, users) => {
```
Comment (thesudeshdas) [inline client/src/api/fakeAuthApi.js:1]: File requires indentation

File: client/src/api/fakeAuthApi.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+           }
+           reject({ success: false,status:401,message:"Invalid password"});
+        },2000)
+    })
+}
+
+export const signup = (email,users) => {
+
+    const findUser = (email) => {
+        return users.find( user => user.email === email)
```
Comment (thesudeshdas) [inline client/src/api/fakeAuthApi.js:25]: Arrow functions returns by default, it can be cleaned here

File: client/src/auth/ForgetPassword.js [hunk 18]
```diff
@@ -0,0 +1,131 @@
+import {useState} from 'react';
```
Comment (thesudeshdas) [inline client/src/auth/ForgetPassword.js:1]: File requires indentation

File: client/src/auth/ForgetPassword.js [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    const handleChange = (e) => {
+       const {value,name} = e.target;
+       setState( state => ({...state,[name]:value}) )
+    }
+
+
+    const formValidate = (state) => {
+        let password,confirmPassword;
+ 
+        if(!/^(?=.*\d)(?=.*[a-z])(?=.*[a-zA-Z]).{8,}$/.test(state.password)){
```
Comment (thesudeshdas) [inline client/src/auth/ForgetPassword.js:30]: Create a variable for regex. That way, in future it'd be easy to make changes. Also, reusing a variable is easy, less chances of errors of using the regex itself.

File: client/src/auth/ForgetPassword.js [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+           // logic code
+           setIsEmail(true)
+           return;
+       }
+       if(!isOtp){
+           // logic code
+           setIsOtp(true)
+           return;
+       }
+    //    if(formValidate(state)){
```
Comment (thesudeshdas) [inline client/src/auth/ForgetPassword.js:57]: Remove unnecessary comments

**Review by thesudeshdas — CHANGES_REQUESTED** [review]

**Pull request merged**